File: BHAVRP_Python/cujae/inf/citi/om/controller/controller.py
from typing import List
from .tools.tools import Tools
from .tools.order_type import OrderType
from ..assignment.assignment import Assignment;
from ..factory.methods.factory_assignment import FactoryAssignment
from ..factory.interfaces.assignment_type import AssignmentType
from ..factory.interfaces.ifactory_assignment import IFactoryAssignment
from ..problem.input.problem import Problem
from ..problem.output.solution import Solution

class Controller:
    _instance = None   # Atributo para la instancia Singleton
    order_type = OrderType.INPUT  # Instancia de OrderType
    distance_type = Assignment.distance_type
    
    def __init__(self):
        if Controller._instance is not None:
            raise Exception("This class is a Singleton! Use get_instance() instead.")
        self.solution = None
        
    # Método Singleton para obtener la única instancia de la clase controladora.  
    @staticmethod
    def get_instance():
        if Controller._instance is None:
            Controller._instance = Controller()
        return Controller._instance

    def get_solution(self) -> Solution:
        return self.solution
    
    def set_solution(self, solution: Solution):
        self.solution = solution
         
    # Método encargado de cargar los datos del problema (incluido las coordenadas) usando listas de distancias
    def load_problem(
        self, 
        id_customers: List[int], 
        request_customers: List[float], 
        axis_x_customers: List[float], 
        axis_y_customers: List[float], 
        id_depots: List[int], 
        axis_x_depots: List[float], 
        axis_y_depots: List[float], 
        count_vehicles: List[List[int]], 
        capacity_vehicles: List[List[float]]
    ) -> bool:
        loaded: bool = False
        
        print("ENTRADA A LA CARGA DE DATOS")
        print("-------------------------------------------------------------------------------")
        print(f"CANTIDAD DE CLIENTES: {len(id_customers)}")
        print("-------------------------------------------------------------------------------")
        
        for i in range(len(id_customers)):
            print(f"ID CLIENTE: {id_customers[i]}")
            print(f"DEMANDA : {request_customers[i]}")
            print(f"X : {axis_x_customers[i]}")
            print(f"Y : {axis_y_customers[i]}")

        print(f"CANTIDAD DE DEPÓSITOS: {len(id_depots)}")
        print("-------------------------------------------------------------------------------")

        total_capacity = 0.0

        for i in range(len(id_depots)):
            print(f"ID DEPÓSITO: {id_depots[i]}")
            print(f"X : {axis_x_depots[i]}")
            print(f"Y : {axis_y_depots[i]}")

            print(f"CANTIDAD DE FLOTAS DEL DEPÓSITO: {len(count_vehicles[i])}")
            
            total_depot_capacity = 0.0
            
            for j in range(len(count_vehicles[i])):
                vehicles = count_vehicles[i][j]
                capacity = capacity_vehicles[i][j]
                total_depot_capacity += vehicles * capacity
                
                print(f"CANTIDAD DE VEHÍCULOS: {vehicles}")
                print(f"CAPACIDAD DE LOS VEHÍCULOS: {capacity}")

            print(f"CAPACIDAD TOTAL DEL DEPÓSITO: {total_depot_capacity}")
            total_capacity += total_depot_capacity
            print("-------------------------------------------------------------------------------")

        # Validar las condiciones necesarias para proceder con la carga
        if (id_customers and request_customers and axis_x_customers and axis_y_customers and
            id_depots and axis_x_depots and axis_y_depots and count_vehicles and
            capacity_vehicles):
            
            Problem.get_problem().load_customer(id_customers, request_customers, axis_x_customers, axis_y_customers)
            Problem.get_problem().load_depot(id_depots, axis_x_depots, axis_y_depots, count_vehicles, capacity_vehicles)         
            loaded = True
            
        print(f"DEMANDA TOTAL DE LOS CLIENTES: {Problem.get_problem().get_total_request()}")
        print(f"CAPACIDAD TOTAL DE LOS DEPÓSITOS: {total_capacity}")
        print("-------------------------------------------------------------------------------")
        print(f"CARGA EXITOSA: {loaded}")
        print("FIN DE LA CARGA DE DATOS")
        print("-------------------------------------------------------------------------------")

        return loaded
    
    # Método encargado de ejecutar la heurística de asignación
    def execute_assignment(self, assignment_type: AssignmentType):
                
        assignment: Assignment = self.new_assignment(assignment_type)
        
        print("EJECUCIÓN DE LA HEURÍSTICA")
        print("-------------------------------------------------------------------------------")
        print(f"HEURÍSTICA: {assignment_type.name}")
        print("-------------------------------------------------------------------------------")
        
        if assignment_type in [
            AssignmentType.NearestByCustomer,
            AssignmentType.SequentialCyclic,
            AssignmentType.CyclicAssignment,
            AssignmentType.Kmeans,
            AssignmentType.CoefficientPropagation,
            AssignmentType.NearestByDepot
        ]:
            if self.order_type == OrderType.ASCENDENT:
                Tools.ascendent_ordenate()
            elif self.order_type == OrderType.DESCENDENT or self.order_type == OrderType.INPUT:
                Tools.descendent_ordenate()
            elif self.order_type == OrderType.RANDOM:
                Tools.random_ordenate()
        
        solution: Solution = assignment.to_clustering()
        
        print("-------------------------------------------------------------------------------")
        print("SOLUTION:")
        print(f"CANTIDAD DE CLUSTERS: {len(solution.get_clusters())}")
        print("-------------------------------------------------------------------------------")
        
        for cluster in solution.get_clusters():
            print(f"ID CLUSTER: {cluster.get_id_cluster()}")
            print(f"DEMANDA DEL CLUSTER: {cluster.get_request_cluster()}")
            print(f"TOTAL DE ELEMENTOS DEL CLUSTER: {len(cluster.get_items_of_cluster())}")
            print("-------------------------------------------------------------------------------")
            
            for item in cluster.get_items_of_cluster():
                print(f"ID DEL ELEMENTO: {int(item)}")
            print("-------------------------------------------------------------------------------")
        
        print(f"TOTAL DE CLIENTES NO ASIGNADOS: {solution.get_total_unassigned_items()}")
        
        if solution.get_total_unassigned_items() > 0:
            print("CLIENTES NO ASIGNADOS:")
            print("-------------------------------------------------------------------------------")
            
            for item in solution.get_unassigned_items():
                print(f"ID DEL ELEMENTO NO ASIGNADO: {int(item)}")
            print("-------------------------------------------------------------------------------")
        
        id_dep_without_cust = self.get_depots_without_customers(solution)
        
        print(f"TOTAL DE DEPÓSITOS SIN ASIGNACIÓN DE CLIENTES: {len(id_dep_without_cust)}")
        
        if id_dep_without_cust:
            print("DEPÓSITOS SIN ASIGNACIÓN DE CLIENTES:")
            print("-------------------------------------------------------------------------------")
            
            for depot_id in id_dep_without_cust:
                print(f"ID DEL DEPÓSITO: {int(depot_id)}")
            print("-------------------------------------------------------------------------------")
        
        # self.clean_controller()
        
    # Método encargado de crear una método de asignación
    def new_assignment(self, type_assignment: AssignmentType) -> Assignment:
        i_factory_assignment: IFactoryAssignment = FactoryAssignment()
        assignment = i_factory_assignment.create_assignment(type_assignment)
        return assignment

    # Método encargado de devolver la demanda cubierta para un depósito dado en la solución
    def request_for_depot(self, id_depot: int) -> float:
        request_depot: float = 0.0

        i = 0
        found: bool = False

        while i < len(self.solution.get_clusters()) and not found:
            if self.solution.get_clusters()[i].get_id_cluster() == id_depot:
                request_depot = self.solution.get_clusters()[i].get_request_cluster()
                found = True
            else:
                i += 1

        return request_depot
    
    # Método encargado de devolver los depósitos a los que no se les asigno ningún cliente en la solución
    def get_depots_without_customers(self, solution: Solution) -> List[int]:
        id_depots: List[int] = []

        total_depots = len(Problem.get_problem().get_depots())
        total_clusters = len(solution.get_clusters())

        if total_clusters < total_depots:
            for i in range(total_depots):
                j = 0
                found = False

                id_depot = Problem.get_problem().get_depots()[i].get_id_depot()

                while j < total_clusters and not found:
                    if solution.get_clusters()[j].get_id_cluster() == id_depot:
                        found = True
                    else:
                        j += 1

                if not found:
                    id_depots.append(id_depot)

        return id_depots

    # Método encargado de restaurar los parámetros globales de la clase Controller
    def clean_controller(self):
        self.solution.get_clusters().clear()
        self.solution.get_unassigned_items().clear()
        self.solution = None
    
    # Método encargado de destruir la instancia de la controladora
    @staticmethod
    def destroy_controller():
        Controller._instance = None

File: BHAVRP_Java/BHAVRPv2.0/BHAVRP2.0/src/cujae/inf/ic/om/controller/Controller.java
package cujae.inf.ic.om.controller;

import java.lang.reflect.InvocationTargetException;

import java.util.ArrayList;

import cujae.inf.ic.om.assignment.Assignment;

import cujae.inf.ic.om.controller.tools.OrderType;
import cujae.inf.ic.om.controller.tools.Tools;

import cujae.inf.ic.om.factory.interfaces.AssignmentType;
import cujae.inf.ic.om.factory.interfaces.IFactoryAssignment;
import cujae.inf.ic.om.factory.methods.FactoryAssignment;

import cujae.inf.ic.om.problem.input.Problem;
import cujae.inf.ic.om.problem.solution.Solution;

public class Controller {
	private static Controller controller = null;
	public static OrderType orderType = OrderType.Input;
	private Solution solution;

	private Controller() {
		super();
	}
	
	/* Método que implementa el Patrón Singleton.*/
	public static Controller getController(){
		if (controller == null)
			controller = new Controller();
		
		return controller;
	}

	public Solution getSolution() {
		return solution;
	}

	public void setSolution(Solution solution) {
		this.solution = solution;
	}
	
	/* Método encargado de cargar los datos del problema (incluido las coordenadas).*/
	public boolean loadProblem(ArrayList<Integer> idCustomers, ArrayList<Double> requestCustomers, ArrayList<Double> axisXCustomers, ArrayList<Double> axisYCustomers, ArrayList<Integer> idDepots, ArrayList<Double> axisXDepots, ArrayList<Double> axisYDepots, ArrayList<ArrayList<Integer>> countVehicles, ArrayList<ArrayList<Double>> capacityVehicles)throws IllegalArgumentException, SecurityException, ClassNotFoundException, InstantiationException, IllegalAccessException, InvocationTargetException, NoSuchMethodException{
	boolean loaded = false;

	System.out.println("ENTRADA A LA CARGA DE DATOS");
	System.out.println("-------------------------------------------------------------------------------");
	System.out.println("CANTIDAD DE CLIENTES: " + idCustomers.size());
	System.out.println("-------------------------------------------------------------------------------");
	for(int i = 0; i < idCustomers.size(); i++)
	{
		System.out.println("ID CLIENTE: " + idCustomers.get(i));
		System.out.println("DEMANDA : " + requestCustomers.get(i));
		System.out.println("X : " + axisXCustomers.get(i));
		System.out.println("Y : " + axisYCustomers.get(i));
	}
	System.out.println("CANTIDAD DE DEPÓSITOS: " + idDepots.size());
	System.out.println("-------------------------------------------------------------------------------");
	
	int totalVehicles = 0;
	double capacityVehicle = 0.0;
	
	for(int i = 0; i < idDepots.size(); i++)
	{
		System.out.println("ID DEPÓSITO: " + idDepots.get(i));
		System.out.println("X : " + axisXDepots.get(i));
		System.out.println("Y : " + axisYDepots.get(i));
		
		System.out.println("CANTIDAD DE FLOTAS DEL DEPÓSITO: " + countVehicles.get(i).size());
		for(int j = 0; j < countVehicles.get(i).size(); j++)
		{
			totalVehicles = countVehicles.get(i).get(j);
			capacityVehicle = capacityVehicles.get(i).get(j);
			
			System.out.println("CANTIDAD DE VEHÍCULOS: " + countVehicles.get(i).get(j));
			System.out.println("CAPACIDAD DE LOS VEHÍCULOS: " + capacityVehicles.get(i).get(j));
		}
		
	//	System.out.println("CAPACIDAD TOTAL DEL DEPÓSITO: " + Problem.getProblem().getTotalCapacityByDepot(idDepots.get(i)));
		System.out.println("CAPACIDAD TOTAL DEL DEPÓSITO: " + (totalVehicles * capacityVehicle));
		
		System.out.println("-------------------------------------------------------------------------------");
	}		
		
		if((idCustomers != null && !idCustomers.isEmpty()) && (requestCustomers != null && !requestCustomers.isEmpty()) && (axisXCustomers != null && !axisXCustomers.isEmpty()) && (axisYCustomers != null && !axisYCustomers.isEmpty()) && 
				(idDepots != null && !idDepots.isEmpty()) && (axisXDepots != null && !axisXDepots.isEmpty()) && (axisYDepots!= null && !axisYDepots.isEmpty()) && (countVehicles != null && !countVehicles.isEmpty()) && 
				(capacityVehicles != null && !capacityVehicles.isEmpty()))
		
		{
			Problem.getProblem().loadCustomer(idCustomers, requestCustomers, axisXCustomers, axisYCustomers);
			Problem.getProblem().loadDepot(idDepots, axisXDepots, axisYDepots, countVehicles, capacityVehicles);
			loaded = true;
		}

		System.out.println("-------------------------------------------------------------------------------");
		System.out.println("DEMANDA TOTAL DE LOS CLIENTES: " + Problem.getProblem().getTotalRequest());
		System.out.println("CAPACIDAD TOTAL DE LOS DEPÓSITOS: " + Problem.getProblem().getTotalCapacity());
		System.out.println("-------------------------------------------------------------------------------");
		System.out.println("CARGA EXITOSA: " + loaded);
		System.out.println("FIN DE LA CARGA DE DATOS");
		System.out.println("-------------------------------------------------------------------------------");
		
		return loaded;
	}
	
	/* Método encargado de ejecutar la heurística de asignación.*/
	public void executeAssignment(AssignmentType assignmentType) throws IllegalArgumentException, SecurityException, ClassNotFoundException, InstantiationException, IllegalAccessException, InvocationTargetException, NoSuchMethodException{
		Assignment assignment = newAssignment(assignmentType);
		
		System.out.println("EJECUCIÓN DE LA HEURÍSTICA");
		System.out.println("-------------------------------------------------------------------------------");
		System.out.println("HEURÍSTICA: " + assignmentType.name());
		System.out.println("-------------------------------------------------------------------------------");
			
		if(assignmentType.equals(AssignmentType.NearestByCustomer) || assignmentType.equals(AssignmentType.SequentialCyclic) || assignmentType.equals(AssignmentType.CyclicAssignment) || assignmentType.equals(AssignmentType.KMEANS) ||
		   assignmentType.equals(AssignmentType.CoefficientPropagation) || assignmentType.equals(AssignmentType.NearestByDepot))
		{ 
			switch(orderType.ordinal())
			{
				case 0:
				{
					Tools.ascendentOrdenate();
					break;
				}
				case 1:
				{
					Tools.descendentOrdenate();
					break;
				}
				case 4:
				{
					Tools.randomOrdenate();
					break;
				}
			}
		}

		solution = assignment.toClustering();
		
		System.out.println("-------------------------------------------------------------------------------");
		System.out.println("SOLUTION: ");
		System.out.println("CANTIDAD DE CLUSTERS: " + solution.getClusters().size());
		System.out.println("-------------------------------------------------------------------------------");
		for(int i = 0; i < solution.getClusters().size(); i++)
		{
			System.out.println("ID CLUSTER: " + solution.getClusters().get(i).getIDCluster());
			System.out.println("DEMANDA DEL CLUSTER: " + solution.getClusters().get(i).getRequestCluster());
			System.out.println("TOTAL DE ELEMENTOS DEL CLUSTER : " + solution.getClusters().get(i).getItemsOfCluster().size());
			System.out.println("-------------------------------------------------------------------------------");
			for(int j = 0; j < solution.getClusters().get(i).getItemsOfCluster().size(); j++)
			{
				System.out.println("ID DEL ELEMENTO: " + solution.getClusters().get(i).getItemsOfCluster().get(j).intValue());
			}	
			System.out.println("-------------------------------------------------------------------------------");
		}
		System.out.println("TOTAL DE CLIENTES NO ASIGNADOS: " + solution.getTotalUnassignedItems());
		if(solution.getTotalUnassignedItems() > 0)
		{
			System.out.println("CLIENTES NO ASIGNADOS: ");
			System.out.println("-------------------------------------------------------------------------------");
			for(int i = 0; i < solution.getUnassignedItems().size(); i++)
				System.out.println("ID DEL ELEMENTO NO ASIGNADO: " + solution.getUnassignedItems().get(i).intValue());
			System.out.println("-------------------------------------------------------------------------------");
		}
		ArrayList<Integer> idDepWithOutCust = new ArrayList<Integer>();
		idDepWithOutCust = getDepotsWithOutCustomers(); // here
		System.out.println("TOTAL DE DEPÓSITOS SIN ASIGNACIÓN DE CLIENTES: " + idDepWithOutCust.size());
		if(!idDepWithOutCust.isEmpty())
		{
			System.out.println("DEPÓSITOS SIN ASIGNACIÓN DE CLIENTES: ");
			System.out.println("-------------------------------------------------------------------------------");
			for(int i = 0; i < idDepWithOutCust.size(); i++)
				System.out.println("ID DEL DEPÓSITO: " + idDepWithOutCust.get(i).intValue());
		}
		
		//cleanController();
	}
	
	/* Método encargado de crear una método de asignación.*/
	private Assignment newAssignment(AssignmentType typeAssignment) throws IllegalArgumentException, SecurityException, ClassNotFoundException, InstantiationException, IllegalAccessException, InvocationTargetException, NoSuchMethodException {
		IFactoryAssignment iFactoryAssignment = new FactoryAssignment();
		Assignment assignment = iFactoryAssignment.createAssignment(typeAssignment);
		return assignment;
	}
	
	/*Método encargado de devolver los depósitos a los que no se les asigno ningún cliente en la solución.*/
	private ArrayList<Integer> getDepotsWithOutCustomers(){
		ArrayList<Integer> idDepots = new ArrayList<Integer>();
		
		int totalDepots = Problem.getProblem().getDepots().size();
		int totalClusters = solution.getClusters().size();
		
		if(totalClusters < totalDepots)
		{
			for(int i = 0; i < totalDepots; i++)
			{
				int j = 0;
				boolean found = false;
				
				int idDepot = Problem.getProblem().getDepots().get(i).getIDDepot();
				
				while((j < totalClusters) && (!found))
				{
					if(solution.getClusters().get(j).getIDCluster() == idDepot)
						found = true;
					else
						j++;
				}
				
				if(!found)
					idDepots.add(idDepot);
			}
		}	
		return idDepots;
	}
	
	/*Método encargado de restaurar los parámetros globales de la clase Controller.*/
	public void cleanController() {
		solution.getClusters().clear();
		solution.getUnassignedItems().clear();
		solution = null;
	}
	
	/* Método encargado de destruir la instancia de la controladora.*/
	public static void destroyController() {
		controller = null;
	}
}
/*
	// Método encargado de cargar los datos del problema usando listas de distancias y el tipo de distancia
	public boolean loadProblem(ArrayList<Integer> idCustomers, ArrayList<Double> requestCustomers, ArrayList<Integer> idDepots, 
			ArrayList<ArrayList<Integer>> countVehicles, ArrayList<ArrayList<Double>> capacityVehicles,  
			ArrayList<ArrayList<Double>> listDistances)throws IllegalArgumentException, SecurityException, 
			ClassNotFoundException, InstantiationException, IllegalAccessException, InvocationTargetException, NoSuchMethodException {
		boolean loaded = false;

		System.out.println("ENTRADA A LA CARGA DE DATOS");
		System.out.println("-------------------------------------------------------------------------------");
		System.out.println("CANTIDAD DE CLIENTES: " + idCustomers.size());
		System.out.println("-------------------------------------------------------------------------------");
		for(int i = 0; i < idCustomers.size(); i++)
		{
			System.out.println("ID CLIENTE: " + idCustomers.get(i));
			System.out.println("DEMANDA : " + requestCustomers.get(i));
		}
		System.out.println("CANTIDAD DE DEPÓSITOS: " + idDepots.size());
		System.out.println("-------------------------------------------------------------------------------");
		for(int i = 0; i < idDepots.size(); i++)
		{
			System.out.println("ID DEPÓSITO: " + idDepots.get(i));
			//System.out.println("CAPACIDAD DEL DEPÓSITO: " + Problem.getProblem().getTotalCapacityByDepot(depot));
			System.out.println("CANTIDAD DE FLOTAS DEL DEPÓSITO: " + countVehicles.get(i).size());
			for(int j = 0; j < countVehicles.get(i).size(); j++)
			{
				System.out.println("CANTIDAD DE VEHÍCULOS: " + countVehicles.get(i).get(j));
				System.out.println("CAPACIDAD DE LOS VEHÍCULOS: " + capacityVehicles.get(i).get(j));
			}
			System.out.println("-------------------------------------------------------------------------------");
		}
		
		if((idCustomers != null && !idCustomers.isEmpty()) && (requestCustomers != null && !requestCustomers.isEmpty()) && 
				(idCustomers.size() == requestCustomers.size()) && (idDepots != null && !idDepots.isEmpty()) && 
				(countVehicles != null && !countVehicles.isEmpty()) && (capacityVehicles != null && !capacityVehicles.isEmpty()) && 
				(idDepots.size() == countVehicles.size()) && (idDepots.size() == capacityVehicles.size()) && (countVehicles.size() == capacityVehicles.size()) && 
				(listDistances != null && !listDistances.isEmpty()) && (listDistances.size() == (idCustomers.size() + idDepots.size())))
		
		{
			Problem.getProblem().loadCustomer(idCustomers, requestCustomers);
			Problem.getProblem().loadDepot(idDepots, countVehicles, capacityVehicles);

			if((Problem.getProblem().getTotalCapacity() >= Problem.getProblem().getTotalRequest()))
			{
				loaded = true;
				Problem.getProblem().fillCostMatrix(listDistances);
			}
		}
		
		System.out.println("-------------------------------------------------------------------------------");
		System.out.println("DEMANDA TOTAL DE LOS CLIENTES: " + Problem.getProblem().getTotalRequest());
		System.out.println("CAPACIDAD TOTAL DE LOS DEPÓSITOS: " + Problem.getProblem().getTotalCapacity());
		System.out.println("-------------------------------------------------------------------------------");
		System.out.println("CARGA EXITOSA: " + loaded);
		System.out.println("FIN DE LA CARGA DE DATOS");
		System.out.println("-------------------------------------------------------------------------------");
		
		return loaded;
	}

	// Método encargado de cargar los datos del problema usando matriz de costo.
	public boolean loadProblem(ArrayList<Integer> idCustomers, ArrayList<Double> requestCustomers, ArrayList<Integer> idDepots, 
			ArrayList<ArrayList<Integer>> countVehicles, ArrayList<ArrayList<Double>> capacityVehicles, 
			NumericMatrix costMatrix)throws IllegalArgumentException, SecurityException, ClassNotFoundException, 
			InstantiationException, IllegalAccessException, InvocationTargetException, NoSuchMethodException {
		boolean loaded = false;

		System.out.println("ENTRADA A LA CARGA DE DATOS");
		System.out.println("-------------------------------------------------------------------------------");
		System.out.println("CANTIDAD DE CLIENTES: " + idCustomers.size());
		System.out.println("-------------------------------------------------------------------------------");
		for(int i = 0; i < idCustomers.size(); i++)
		{
			System.out.println("ID CLIENTE: " + idCustomers.get(i));
			System.out.println("DEMANDA : " + requestCustomers.get(i));
		}
		System.out.println("CANTIDAD DE DEPÓSITOS: " + idDepots.size());
		System.out.println("-------------------------------------------------------------------------------");
		for(int i = 0; i < idDepots.size(); i++)
		{
			System.out.println("ID DEPÓSITO: " + idDepots.get(i));
			System.out.println("CANTIDAD DE FLOTAS DEL DEPÓSITO: " + countVehicles.get(i).size());
			System.out.println("-------------------------------------------------------------------------------");
			for(int j = 0; j < countVehicles.get(i).size(); i++)
			{
				System.out.println("CANTIDAD DE VEHÍCULOS: " + countVehicles.get(i).get(j));
				System.out.println("CAPACIDAD DE LOS VEHÍCULOS: " + capacityVehicles.get(i).get(j));
			}
			System.out.println("-------------------------------------------------------------------------------");
		}
		
		if((idCustomers != null && !idCustomers.isEmpty()) && (requestCustomers != null && !requestCustomers.isEmpty()) && 
				(idCustomers.size() == requestCustomers.size()) && (idDepots != null && !idDepots.isEmpty()) && 
				(countVehicles != null && !countVehicles.isEmpty()) && (capacityVehicles != null && !capacityVehicles.isEmpty()) && 
				(idDepots.size() == countVehicles.size()) && (idDepots.size() == capacityVehicles.size()) && (countVehicles.size() == capacityVehicles.size()) && 
				(costMatrix.getColCount() == (idCustomers.size() + idDepots.size())) && (costMatrix.getRowCount() == (idCustomers.size() + idDepots.size())))
		{
			Problem.getProblem().loadCustomer(idCustomers, requestCustomers);
			Problem.getProblem().loadDepot(idDepots, countVehicles, capacityVehicles);

			if((Problem.getProblem().getTotalCapacity() >= Problem.getProblem().getTotalRequest()))
			{
				loaded = true;
				Problem.getProblem().setCostMatrix(costMatrix);
			}
		}

		System.out.println("-------------------------------------------------------------------------------");
		System.out.println("DEMANDA TOTAL DE LOS CLIENTES: " + Problem.getProblem().getTotalRequest());
		System.out.println("CAPACIDAD TOTAL DE LOS DEPÓSITOS: " + Problem.getProblem().getTotalCapacity());
		System.out.println("-------------------------------------------------------------------------------");
		System.out.println("CARGA EXITOSA: " + loaded);
		System.out.println("FIN DE LA CARGA DE DATOS");
		System.out.println("-------------------------------------------------------------------------------");
		
		return loaded;
	}
	
	// Método encargado de cargar los datos del problema (incluido las coordenadas) usando listas de distancias
	public boolean loadProblem(ArrayList<Integer> idCustomers, ArrayList<Double> requestCustomers, ArrayList<Double> axisXCustomers, 
			ArrayList<Double> axisYCustomers, ArrayList<Integer> idDepots, ArrayList<Double> axisXDepots, ArrayList<Double> axisYDepots, 
			ArrayList<ArrayList<Integer>> countVehicles, ArrayList<ArrayList<Double>> capacityVehicles, 
			ArrayList<ArrayList<Double>> listDistances)throws IllegalArgumentException, SecurityException, 
			ClassNotFoundException, InstantiationException, IllegalAccessException, InvocationTargetException, NoSuchMethodException {
	boolean loaded = false;

	System.out.println("ENTRADA A LA CARGA DE DATOS");
	System.out.println("-------------------------------------------------------------------------------");
	System.out.println("CANTIDAD DE CLIENTES: " + idCustomers.size());
	System.out.println("-------------------------------------------------------------------------------");
	for(int i = 0; i < idCustomers.size(); i++)
	{
		System.out.println("ID CLIENTE: " + idCustomers.get(i));
		System.out.println("DEMANDA : " + requestCustomers.get(i));
		System.out.println("X : " + axisXCustomers.get(i));
		System.out.println("Y : " + axisYCustomers.get(i));
	}
	System.out.println("CANTIDAD DE DEPÓSITOS: " + idDepots.size());
	System.out.println("-------------------------------------------------------------------------------");
	
	int totalVehicles = 0;
	double capacityVehicle = 0.0;
	
	for(int i = 0; i < idDepots.size(); i++)
	{
		System.out.println("ID DEPÓSITO: " + idDepots.get(i));
		System.out.println("X : " + axisXDepots.get(i));
		System.out.println("Y : " + axisYDepots.get(i));
		
		System.out.println("CANTIDAD DE FLOTAS DEL DEPÓSITO: " + countVehicles.get(i).size());
		for(int j = 0; j < countVehicles.get(i).size(); j++)
		{
			totalVehicles = countVehicles.get(i).get(j);
			capacityVehicle = capacityVehicles.get(i).get(j);
			
			System.out.println("CANTIDAD DE VEHÍCULOS: " + countVehicles.get(i).get(j));
			System.out.println("CAPACIDAD DE LOS VEHÍCULOS: " + capacityVehicles.get(i).get(j));
		}
		
	//	System.out.println("CAPACIDAD TOTAL DEL DEPÓSITO: " + Problem.getProblem().getTotalCapacityByDepot(idDepots.get(i)));
		System.out.println("CAPACIDAD TOTAL DEL DEPÓSITO: " + (totalVehicles * capacityVehicle));
		
		System.out.println("-------------------------------------------------------------------------------");
	}		
		
		if((idCustomers != null && !idCustomers.isEmpty()) && (requestCustomers != null && !requestCustomers.isEmpty()) && (axisXCustomers != null && !axisXCustomers.isEmpty()) && (axisYCustomers != null && !axisYCustomers.isEmpty()) && 
				(idDepots != null && !idDepots.isEmpty()) && (axisXDepots != null && !axisXDepots.isEmpty()) && (axisYDepots!= null && !axisYDepots.isEmpty()) && (countVehicles != null && !countVehicles.isEmpty()) && 
				(capacityVehicles != null && !capacityVehicles.isEmpty()) && (listDistances != null && !listDistances.isEmpty()))
		
		{
			Problem.getProblem().loadCustomer(idCustomers, requestCustomers, axisXCustomers, axisYCustomers);
			Problem.getProblem().loadDepot(idDepots, axisXDepots, axisYDepots, countVehicles, capacityVehicles);
								
			if((Problem.getProblem().getTotalCapacity() >= Problem.getProblem().getTotalRequest()))
			{
				loaded = true;
				Problem.getProblem().fillCostMatrix(listDistances);
			}
		}

		System.out.println("-------------------------------------------------------------------------------");
		System.out.println("DEMANDA TOTAL DE LOS CLIENTES: " + Problem.getProblem().getTotalRequest());
		System.out.println("CAPACIDAD TOTAL DE LOS DEPÓSITOS: " + Problem.getProblem().getTotalCapacity());
		System.out.println("-------------------------------------------------------------------------------");
		System.out.println("CARGA EXITOSA: " + loaded);
		System.out.println("FIN DE LA CARGA DE DATOS");
		System.out.println("-------------------------------------------------------------------------------");
		
		return loaded;
	}
		
	// Método encargado de cargar los datos del problema (incluido las coordenadas) usando listas de distancias
	public boolean loadProblem(ArrayList<Integer> idCustomers, ArrayList<Double> requestCustomers, ArrayList<Double> axisXCustomers, ArrayList<Double> axisYCustomers, ArrayList<Integer> idDepots, ArrayList<Double> axisXDepots, ArrayList<Double> axisYDepots, ArrayList<ArrayList<Integer>> countVehicles, ArrayList<ArrayList<Double>> capacityVehicles,
			NumericMatrix costMatrix)throws IllegalArgumentException, SecurityException, ClassNotFoundException, InstantiationException, IllegalAccessException, InvocationTargetException, NoSuchMethodException{
	boolean loaded = false;

	System.out.println("ENTRADA A LA CARGA DE DATOS");
	System.out.println("-------------------------------------------------------------------------------");
	System.out.println("CANTIDAD DE CLIENTES: " + idCustomers.size());
	System.out.println("-------------------------------------------------------------------------------");
	for(int i = 0; i < idCustomers.size(); i++)
	{
		System.out.println("ID CLIENTE: " + idCustomers.get(i));
		System.out.println("DEMANDA : " + requestCustomers.get(i));
		System.out.println("X : " + axisXCustomers.get(i));
		System.out.println("Y : " + axisYCustomers.get(i));
	}
	System.out.println("CANTIDAD DE DEPÓSITOS: " + idDepots.size());
	System.out.println("-------------------------------------------------------------------------------");
	for(int i = 0; i < idDepots.size(); i++)
	{
		System.out.println("ID DEPÓSITO: " + idDepots.get(i));
		System.out.println("X : " + axisXDepots.get(i));
		System.out.println("Y : " + axisYDepots.get(i));
		//System.out.println("CAPACIDAD DEL DEPÓSITO: " + Problem.getProblem().getTotalCapacityByDepot(depot));
		System.out.println("CANTIDAD DE FLOTAS DEL DEPÓSITO: " + countVehicles.get(i).size());
		for(int j = 0; j < countVehicles.get(i).size(); j++)
		{
			System.out.println("CANTIDAD DE VEHÍCULOS: " + countVehicles.get(i).get(j));
			System.out.println("CAPACIDAD DE LOS VEHÍCULOS: " + capacityVehicles.get(i).get(j));
		}
		System.out.println("-------------------------------------------------------------------------------");
	}		
		
		if((idCustomers != null && !idCustomers.isEmpty()) && (requestCustomers != null && !requestCustomers.isEmpty()) && (axisXCustomers != null && !axisXCustomers.isEmpty()) && (axisYCustomers != null && !axisYCustomers.isEmpty()) && 
				(idDepots != null && !idDepots.isEmpty()) && (axisXDepots != null && !axisXDepots.isEmpty()) && (axisYDepots!= null && !axisYDepots.isEmpty()) && (countVehicles != null && !countVehicles.isEmpty()) && 
				(capacityVehicles != null && !capacityVehicles.isEmpty()) && (costMatrix.getColCount() == (idCustomers.size() + idDepots.size())) && (costMatrix.getRowCount() == (idCustomers.size() + idDepots.size())))
		
		{
			Problem.getProblem().loadCustomer(idCustomers, requestCustomers, axisXCustomers, axisYCustomers);
			Problem.getProblem().loadDepot(idDepots, axisXDepots, axisYDepots, countVehicles, capacityVehicles);
								
			if((Problem.getProblem().getTotalCapacity() >= Problem.getProblem().getTotalRequest()))
			{
				loaded = true;
				Problem.getProblem().setCostMatrix(costMatrix);
			}
		}

		System.out.println("-------------------------------------------------------------------------------");
		System.out.println("DEMANDA TOTAL DE LOS CLIENTES: " + Problem.getProblem().getTotalRequest());
		System.out.println("CAPACIDAD TOTAL DE LOS DEPÓSITOS: " + Problem.getProblem().getTotalCapacity());
		System.out.println("-------------------------------------------------------------------------------");
		System.out.println("CARGA EXITOSA: " + loaded);
		System.out.println("FIN DE LA CARGA DE DATOS");
		System.out.println("-------------------------------------------------------------------------------");
		
		return loaded;
	}
}
*/